### PR TITLE
[Wasi-nn] PyTorch AOTI

### PR DESCRIPTION
**Pull request by LFsWang**
https://pytorch.org/docs/stable/torch.compiler_aot_inductor.html

Add a support to load the AOTInductor for PyTorch.

Note: use AOTInductor build by pip need the wasmedge with `_GLIBCXX_USE_CXX11_ABI=0`

- [x] Pump PyTorch from 1.8.0 to 2.4.1
- [x] Pump PyTorch from 2.4.1 to 2.5.1
- [x] Check old examples work
- [x] Merge this with origin PyTorch plugin

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Comment by LFsWang** [comment]
note: for manylinux_2_28_x86_64 , use  `add_definitions(-D_GLIBCXX_USE_CXX11_ABI=0)` and install pytorch with no c++11 abi will work

**Comment by hydai** [comment]
> note: for manylinux_2_28_x86_64 , use `add_definitions(-D_GLIBCXX_USE_CXX11_ABI=0)` and install pytorch with no c++11 abi will work

I see, so there is still a compatible issue about the C++11 ABI in manylinux_2_28.

File: plugins/wasi_nn/wasinn_torch.cpp [hunk 1]
```diff
@@ -10,6 +10,157 @@
...
 #ifdef WASMEDGE_PLUGIN_WASI_NN_BACKEND_TORCH
+
+Expect<ErrNo> TorchScript::setDevice(Device Device) {
+  if (Device == Device::CPU) {
+    TorchDevice = at::kCPU;
+    return ErrNo::Success;
+  } else if (Device == Device::GPU) {
+    if (!torch::cuda::is_available()) {
+      spdlog::error(
+          "[WASI-NN] CUDA Unavailable, platform Cannot support GPU target.");
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_torch.cpp:21]:
```
"[WASI-NN] Torch: CUDA Unavailable. Please check if the installed Torch version or platform supports CUDA."sv
```

File: plugins/wasi_nn/wasinn_torch.cpp [hunk 2]
```diff
@@ -10,6 +10,157 @@
...
+      return ErrNo::InvalidArgument;
+    }
+    TorchDevice = at::kCUDA;
+    return ErrNo::Success;
+  }
+
+  spdlog::error("[WASI-NN] PyTorch Only support CPU and GPU target.");
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_torch.cpp:28]:
```
"[WASI-NN] Torch: Unknown target device. We currently support only CPU and GPU targets."sv
```

File: plugins/wasi_nn/wasinn_torch.cpp [hunk 3]
```diff
@@ -10,6 +10,157 @@
...
+  } else if (RawOutput.isTuple()) {
+    auto OutTensorsTuple = RawOutput.toTuple()->elements();
+    for (auto &OneOf : OutTensorsTuple) {
+      Out.push_back(OneOf.toTensor().clone());
+    }
+  } else if (RawOutput.isTensor()) {
+    auto OutTensor = RawOutput.toTensor();
+    Out.push_back(OutTensor.clone());
+  } else {
+    spdlog::error("[WASI-NN] PyTorch backend only supports output a tensor, "
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_torch.cpp:71]:
```
"[WASI-NN] Torch: The output can only be one of the following tensor types: a tensor, a list of tensors, or a tuple of tensors."sv
```

**Review by hydai — CHANGES_REQUESTED** [review]
Let's use the `[WASI-NN] Torch:` prefix for the logs.

File: plugins/wasi_nn/wasinn_torch.cpp [hunk 4]
```diff
@@ -10,11 +10,167 @@
...
+    return ErrNo::InvalidArgument;
+#endif
+  }
+
+  spdlog::error("[WASI-NN] Torch: Unknown target device. We currently support "
+                "only CPU and GPU targets."sv);
+  return ErrNo::InvalidArgument;
+}
+
+Expect<ErrNo> AOTInductor::loadFromBiary(std::istream &, Device) {
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_torch.cpp:109]: `Binary`

File: plugins/wasi_nn/wasinn_torch.cpp [hunk 5]
```diff
@@ -10,11 +10,167 @@
...
+                               std::vector<at::Tensor> &Out) {
+  std::vector<at::Tensor> RawOutput = TorchModel->run(In);
+
+  for (auto &OneOf : RawOutput) {
+    Out.push_back(OneOf.clone());
+  }
+  return ErrNo::Success;
+}
+
+PyModelBackend GuessPyModelBackendType(const std::string_view &Model) {
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_torch.cpp:149]: `guessPyModelBackendType`

File: plugins/wasi_nn/wasinn_torch.h [hunk 6]
```diff
@@ -19,16 +24,65 @@ struct WasiNNEnvironment;
...
+protected:
   torch::DeviceType TorchDevice = at::kCPU;
 };
 
+class TorchScript : public PyBaseModule {
+  Expect<ErrNo> setDevice(Device Device) override;
+
+public:
+  Expect<ErrNo> loadFromPath(const std::string &Path, Device Device) override;
+  Expect<ErrNo> loadFromBiary(std::istream &In, Device Device) override;
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_torch.h:49]: `Binary`

File: plugins/wasi_nn/wasinn_torch.h [hunk 7]
```diff
@@ -19,16 +24,65 @@ struct WasiNNEnvironment;
...
+  torch::jit::Module TorchModel;
+};
+
+class AOTInductor : public PyBaseModule {
+  Expect<ErrNo> setDevice(Device Device) override;
+
+public:
+  AOTInductor();
+  Expect<ErrNo> loadFromPath(const std::string &Path, Device Device) override;
+  Expect<ErrNo> loadFromBiary(std::istream &In, Device Device) override;
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_torch.h:62]: `Binary`

**Review by hydai — CHANGES_REQUESTED** [review]

**Review by hydai — APPROVED** [review]

**Pull request merged**